boards/xtensa/esp32/common/src/esp32_board_spiflash.c

```diff
@@ -405,6 +405,17 @@ static int init_storage_partition(void)
       return ret;
     }
 
+#elif defined(CONFIG_MTD_NVBLK)
+
+  ret = nvblk_initialize(0, mtd, CONFIG_MTD_NVBLK_DEFAULT_LBS,
```

Comment:
it's better to call register_mtddriver and hook nvblk in the common place:
https://github.com/apache/nuttx/blob/master/fs/driver/fs_mtdproxy.c#L168
so, we can wrap mtd with either ftl/nvblk/dhara automatically.

Reply:
Can the automatic detection work if, let's say, both `CONFIG_FTL` and `CONFIG_MTD_NVBLK` (or dhara) are enabled? I don't think there is a way to force the interface from an application layer (and likely there shouldn't be one).

Reply:
> Can the automatic detection work if, let's say, both `CONFIG_FTL` and `CONFIG_MTD_NVBLK` (or dhara) are enabled?

I assume that only one FTL will be selected in one project.

> I don't think there is a way to force the interface from an application layer (and likely there shouldn't be one).

if the multiple FTL exist, mtd_proxy could add a choice option to select simple/dhara/nvblk. If the default isn't suitable, ftl_initialize, nvblk_initialize or dhara_initialize need be called manually.


> I was thinking the same as @michallenc, but maybe it is possible. Also how to pass setup parameters like logical block size (lbs), io block size (iobs) and number of spare erase blocks (speb) in an automatic detection?

apply the default value from Kconfig.

Reply:
@xiaoxiang781216 multiple ftl support should be possible. Adding a choice option for mtd_proxy could work, but is also very limiting: how to support mtd_proxy with FTL and dhara/nvblk. FTL, dhara and nvblk are libraries that should be coupled with a specific MTD, what library is used cannot be queried from the MTD structure. It is simpler to let the user define the type of block device (ftl, dhara, nvblk, ...).

Reply:
>>  Can the automatic detection work if, let's say, both CONFIG_FTL and CONFIG_MTD_NVBLK (or dhara) are enabled?
> 
> I assume that only one FTL will be selected in one project.

It should be possible to use more that one. Typical use case is if you have two flashes, internal and external. You might want simple FTL on internal for firmare storage, updates, etc and some wear leveling on external for configuration, logs and so on.

Reply:
> > > Can the automatic detection work if, let's say, both CONFIG_FTL and CONFIG_MTD_NVBLK (or dhara) are enabled?
> > 
> > 
> > I assume that only one FTL will be selected in one project.
> 
> It should be possible to use more that one. Typical use case is if you have two flashes, internal and external. You might want simple FTL on internal for firmare storage, updates, etc and some wear leveling on external for configuration, logs and so on.

BCH/FTL auto wrapper need select one of algo as default, which is used for all mtd instance created by register_mtddriver, the mtd device which has the special requirement, need wrap the algo manually and call register_blockdriver instead.

Reply:
> @xiaoxiang781216 multiple ftl support should be possible. Adding a choice option for mtd_proxy could work, but is also very limiting: how to support mtd_proxy with FTL and dhara/nvblk. FTL, dhara and nvblk are libraries that should be coupled with a specific MTD, what library is used cannot be queried from the MTD structure. It is simpler to let the user define the type of block device (ftl, dhara, nvblk, ...).

The support of multiple algo is still exist, but you need wrap mtd with algo by self in board initialization, auto FTL wrapper need be done by selecting one.

Reply:
> @xiaoxiang781216 with the mtd_proxy solution the block device name seems to be `/dev/tmpbxxx`. How to know the correct block device if it is needed to `dd` into the block device ?

the end user doesn't use block device too, it use /dev/mtdblock0 which is char driver(BCH wrapper).

Reply:
no problem.